manual/modules/ROOT/pages/configuration_defaults.adoc
= Configuration Defaults

== Weather Routing Plugin

This lists *default Configuration* settings. Hitting *Reset all Advanced* will reset the _Advanced Tab_ configuration to these default settings.

NOTE: Upoin intial installation (following the instructions), the plugin should be setup to allow the user to select any of the _Configurations_ from the Main Menu and then press _Compute_ and the routing should _Complete_  

NOTE: If you are _learning_ at this point and want to run the *Examples*, don't change the _configuration defaults_.

== Basic Tab

. All of these settings are set "out of the box" for the first _Compute_.
. Don't tough anything.

=== Start Position 

. Make Weather Route Position by right-click in Opencpn on the chart and select "Weather Route position"
. Select Start Position by selection from the dropdown.

=== Start Time

. Select Start Date & Time -User Selection
. Must within the grib file time window unless climatology is working too.
. Two options _Grib Time_ and _Current Time_

=== End Position

. Make Weather Route Position by right-click in Opencpn on the chart and select "Weather Route position"
. Select Start Position by selection from the dropdown.

=== Time Step

. Intially this is set to 12 hours to have a fast calculation.
. Select between 30 min and 12 hours.
. For routing failures first try setting this for a shorter time step.

=== Boat

. Boat.xml path  - For startup examples, select _plugins\weather_routing\boats\Boat-Climatology.xml_
. Boat.xml edit  - View and get familiar with this interface.
.. In _Boat-Climatology.xml_ 4 Sample Polar files have been added.
.. Highlight one of the Polars and select from the Tabs (Plot, Cross Over and Stats)
.. At the bottom select how to view the polar True/Apparent, etc.
.. In the Right menu select _Edit_ to see how Editing a polar works.
.. Close or Save this menu.

=== Constraints - Defaults

. Max Diverted Course 76 degrees
. Max True Wind 100 knots
. Max Apparent Wind 100 knots
. Max Swell 20 meters

=== Options - Defaults

. Time step: 1 hour 0 min 0 sec
. Detect Land: Checked (when first starting, uncheck it, requires High Res GSHHS)
. Detect Boundary: Unchecked
. Currents: Unchecked
. Optimize tacking: Unchecked

=== Data Source - Defaults

. Default Examples
.. The wind data source used for these examples is Climatology.
.. Climatology must be installed and enabled.

. For your own use
.. Download a fresh grib of the entire route area (a little
larger than then route), making sure that it is for a big enough time
interval. 
.. Then set the Grib_pi at least one frame later than the start time.
.. Grib Unchecked - You will need to check this.
.. Climatology - Leave disabled or enabled depending..
... Use Climatology alone.
... Use Climatology (30 year Noaa data averages) to extend long trip calcs both at the start and finish. 
... When data sources change there will be a color indication.
. Last Valid if Data Deficient - Uncheck. If your grib file is not long enough checking this can help.

== Advanced Tab

. Generally the default constraints work well.
. These constraints speed up calculations.
. Some adjustments may be necessary if a route does not complete.
. However first decrease _Time Step_ and increase _Max Diverted Course_

=== Constraints - Defaults

. Max Latitude: 90 degrees
. Wind vs Current: 0 (disabled) Factor for waves and hobby horsing
. Max Course Angle: 180 degrees
. Max Search Angle: 120 degrees
. Cyclones Avoid Crossing Cyclone Tracks: Unchecked  (Climatology needed)
.. within X months X  days: within 1 month 0 days

=== Options - Defaults

. Inverted Regions: Unchecked
. Integrator: Newton
. Anchoring: Unchecked
. Wind Strength: 100 percent
. Tacking Time: 0 (disabled)
. Safety Margin from land: 0 nm
. Courses Relative to True Wind: 
.. From 0 degrees to 180 degrees
.. by 5 degrees steps

If you are having trouble, set the Configuration to these defaults or
select *Configuration > Reset all Advanced*:



manual/weather_routing/modules/ROOT/pages/settings.adoc
== Configuration Defaults

=== Weather Routing Plugin

The following is a list of the default *Configuration* Constraints and
Options. Use this list to check your settings if you are having trouble.

=== Basic Tab
==== Start
==== End
==== Date
==== Time
==== Boat

. Boat.xml path
. Boat.xml edit

==== Constraints

. Max Diverted Course 76 degrees
. Max True Wind 100 knots
. Max Apparent Wind 100 knots
. Max Swell 20 meters - Ordered List Item

==== Options

. Time step 1 hour 0 min 0 sec
. Detect Land Checked (when first starting, uncheck it)
. Detect Boundary Unchecked
. Currents Unchecked
. Optimize tacking Unchecked

==== Data Source

. First use a freshly downloaded grib of the entire route area (a little
larger than then route), making sure that it is for a big enough time
interval. Then set the Grib_pi at least one frame later than the start
time.
. Grib Unchecked - You will need to check this.
. Climatology Disabled - You can leave this disabled when you are
starting. You can run routes using Climatology alone, or you can use the
30 year Noaa data averages to extend calculations for a long trip both
at the start and finish. There will be an indicator on the preferred
route when the data source changes.
. Last Valid if Data Deficient - Unchecked If your grib file is not long
enough checking this can help.

=== Advanced Tab

==== Constraints

. Max Latitude 90 degrees
. Wind vs Current 0 (disabled) Factor for waves and hobby horsing
. Max Course Angle 180 degrees
. Max Search Angle 120 degrees
. Cyclones Avoid Crossing Cyclone Tracks Unchecked within X months X
days 1 month 0 days

==== Options

. Inverted Regions Unchecked
. Integrator Newton
. Anchoring Unchecked
. Wind Strength is 100 percent
. Tacking Time 0 (disabled)
. Safety Margin from land 0 nm
. Courses Relative to True Wind from 0 degrees to 180 degrees by 5
degrees

If you are having trouble, set the Configuration to these defaults or
select Configuration > *Reset All Advanced*:

//*Boat* //

debian/changelog
opencpn-plugin-testplugin (1.0-0) unstable; urgency=medium

  * Initial release for O5 using CI

 -- Jon Gough <jon.gough@eclipsesystems.com.au>  Sun, 10 Feb 2019 09:01:17 -0300


src/Configuration.cpp
/***************************************************************************
 *
 * Project:  OpenCPN Weather Routing plugin
 * Author:   Sean D'Epagnier
 *
 ***************************************************************************
 *   Copyright (C) 2013 by Sean D'Epagnier                                 *
 *   sean@depagnier.com                                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,  USA.         *
 ***************************************************************************
 *
 */

#include <wx/wx.h>

#include <stdlib.h>
#include <math.h>
#include <time.h>

#include "Utilities.h"
#include "Boat.h"
#include "RouteMap.h"
#include "ConfigurationDialog.h"
#include "weather_routing_pi.h"

ConfigurationDialog::ConfigurationDialog( wxWindow *parent )
    : ConfigurationDialogBase(parent)
{
    Load();
}

ConfigurationDialog::~ConfigurationDialog( )
{
}

void ConfigurationDialog::Load()
{
    wxFileConfig *pConf = GetOCPNConfigObject();
    pConf->SetPath ( _T( "/PlugIns/WeatherRouting" ) );

    wxString degreesteps;
    m_lDegreeSteps->Clear();
    pConf->Read( _T("DegreeSteps"), &degreesteps, _T("\
40;50;60;70;80;90;105;120;135;150;165;195;210;225;240;255;270;280;290;300;310;320;"));
    while(degreesteps.size()) {
        m_lDegreeSteps->Append(degreesteps.BeforeFirst(';'));
        degreesteps = degreesteps.AfterFirst(';');
    }
 
    int maxdivertedcourse;
    pConf->Read( _T("MaxDivertedCourse"), &maxdivertedcourse, 100);
    m_sMaxDivertedCourse->SetValue(maxdivertedcourse);

    int maxwindknots;
    pConf->Read( _T("MaxWindKnots"), &maxwindknots, 100);
    m_sMaxWindKnots->SetValue(maxwindknots);

    int maxswellmeters;
    pConf->Read( _T("MaxSwellMeters"), &maxswellmeters, 20);
    m_sMaxSwellMeters->SetValue(maxswellmeters);

    int maxlatitude;
    pConf->Read( _T("MaxLatitude"), &maxlatitude, 90);
    m_sMaxLatitude->SetValue(maxlatitude);

    int tackingtime;
    pConf->Read( _T("TackingTime"), &tackingtime, 0);
    m_sTackingTime->SetValue(tackingtime);

    bool usegrib;
    pConf->Read( _T("UseGrib"), &usegrib, false);
    m_cbUseGrib->SetValue(usegrib);

    bool useclimatology;
    pConf->Read( _T("UseClimatology"), &useclimatology, false);
    m_cbUseClimatology->SetValue(useclimatology);

    bool allowdatadeficient;
    pConf->Read( _T("AllowDataDeficient"), &allowdatadeficient, false);
    m_cbAllowDataDeficient->SetValue(allowdatadeficient);

    bool detectland;
    pConf->Read( _T("DetectLand"), &detectland, true);
    m_cbDetectLand->SetValue(detectland);

    bool currents;
    pConf->Read( _T("Currents"), &currents, true);
    m_cbCurrents->SetValue(currents);

    bool invertedregions;
    pConf->Read( _T("InvertedRegions"), &invertedregions, false);
    m_cbInvertedRegions->SetValue(invertedregions);

    bool anchoring;
    pConf->Read( _T("Anchoring"), &anchoring, false);
    m_cbAnchoring->SetValue(anchoring);

    int timestephours;
    pConf->Read( _T("TimeStepHours"), &timestephours, 1);
    m_sTimeStepHours->SetValue(timestephours);

    int timestepminutes;
    pConf->Read( _T("TimeStepMinutes"), &timestepminutes, 0);
    m_sTimeStepMinutes->SetValue(timestepminutes);

    int timestepseconds;
    pConf->Read( _T("TimeStepSeconds"), &timestepseconds, 0);
    m_sTimeStepSeconds->SetValue(timestepseconds);

    wxPoint p = GetPosition();
    pConf->Read ( _T ( "ConfigurationDialogX" ), &p.x, p.x);
    pConf->Read ( _T ( "ConfigurationDialogY" ), &p.y, p.y);
    SetPosition(p);
}

void ConfigurationDialog::Save( )
{
    wxFileConfig *pConf = GetOCPNConfigObject();
    pConf->SetPath ( _T( "/PlugIns/WeatherRouting" ) );

    wxString degreesteps;
    for(unsigned int i=0; i<m_lDegreeSteps->GetCount(); i++)
        degreesteps += m_lDegreeSteps->GetString(i) + _(";");
    pConf->Write( _T("DegreeSteps"), degreesteps);

    pConf->Write( _T("MaxDivertedCourse"), m_sMaxDivertedCourse->GetValue());
    pConf->Write( _T("MaxWindKnots"), m_sMaxWindKnots->GetValue());
    pConf->Write( _T("MaxSwellMeters"), m_sMaxSwellMeters->GetValue());
    pConf->Write( _T("MaxLatitude"), m_sMaxLatitude->GetValue());
    pConf->Write( _T("TackingTime"), m_sTackingTime->GetValue());

    pConf->Write( _T("UseGrib"), m_cbUseGrib->GetValue());
    pConf->Write( _T("UseClimatology"), m_cbUseClimatology->GetValue());
    pConf->Write( _T("AllowDataDeficient"), m_cbAllowDataDeficient->GetValue());

    pConf->Write( _T("DetectLand"), m_cbDetectLand->GetValue());
    pConf->Write( _T("Currents"), m_cbCurrents->GetValue());
    pConf->Write( _T("InvertedRegions"), m_cbInvertedRegions->GetValue());
    pConf->Write( _T("Anchoring"), m_cbAnchoring->GetValue());

    pConf->Write( _T("TimeStepHours"), m_sTimeStepHours->GetValue());
    pConf->Write( _T("TimeStepMinutes"), m_sTimeStepMinutes->GetValue());
    pConf->Write( _T("TimeStepSeconds"), m_sTimeStepSeconds->GetValue());

    wxPoint p = GetPosition();
    pConf->Write ( _T ( "ConfigurationDialogX" ), p.x);
    pConf->Write ( _T ( "ConfigurationDialogY" ), p.y);
}

void ConfigurationDialog::UpdateOptions(RouteMapOptions &options)
{
    options.dt = 60*(60*m_sTimeStepHours->GetValue()
                     + m_sTimeStepMinutes->GetValue())
        + m_sTimeStepSeconds->GetValue();

    if(m_lDegreeSteps->GetCount() < 4) {
        wxMessageDialog mdlg(this, _("Warning: less than 4 different degree steps specified\n"),
                             wxString(_("Weather Routing"), wxOK | wxICON_WARNING));
        mdlg.ShowModal();
    }

    options.DegreeSteps.clear();
    for(unsigned int i=0; i<m_lDegreeSteps->GetCount(); i++) {
        double step;
        m_lDegreeSteps->GetString(i).ToDouble(&step);
        options.DegreeSteps.push_back(positive_degrees(step));
    }
    options.DegreeSteps.sort();

    options.MaxDivertedCourse = m_sMaxDivertedCourse->GetValue();
    options.MaxWindKnots= m_sMaxWindKnots->GetValue();
    options.MaxSwellMeters = m_sMaxSwellMeters->GetValue();
    options.MaxLatitude = m_sMaxLatitude->GetValue();
    options.TackingTime = m_sTackingTime->GetValue();


    options.DetectLand = m_cbDetectLand->GetValue();
    options.Currents = m_cbCurrents->GetValue();
    options.InvertedRegions = m_cbInvertedRegions->GetValue();
    options.Anchoring = m_cbAnchoring->GetValue();

    options.AllowDataDeficient = m_cbAllowDataDeficient->GetValue();
}

void ConfigurationDialog::OnAddDegreeStep( wxCommandEvent& event )
{
    int s = m_lDegreeSteps->GetSelection();
    if(s == -1)
        s = m_lDegreeSteps->GetCount();

    double step;
    m_tDegreeStep->GetValue().ToDouble(&step);
    m_lDegreeSteps->Insert(wxString::Format(_T("%f"), step), s);
    m_tDegreeStep->Clear();
}

void ConfigurationDialog::OnRemoveDegreeStep( wxCommandEvent& event )
{
    int s = m_lDegreeSteps->GetSelection();
    if(s == -1)
        return;

    m_lDegreeSteps->Delete(s);
    m_lDegreeSteps->SetSelection(s);
}

void ConfigurationDialog::OnGenerateDegreeSteps( wxCommandEvent& event )
{
    double value, v;
    m_tDegreeStep->GetValue().ToDouble(&value);

    if(value <= 0 || value > 180) {
        wxMessageDialog mdlg(this, _("Invalid step size, nothing will be done."),
                             wxString(_("Weather Routing"), wxOK | wxICON_WARNING));
        mdlg.ShowModal();
        return;
    }

    m_lDegreeSteps->Clear();
    for(v = value; v < 360; v+=value)
        m_lDegreeSteps->Append(wxString::Format(_T("%.1f"), v));
}

void ConfigurationDialog::OnClearDegreeSteps( wxCommandEvent& event )
{
    m_lDegreeSteps->Clear();
}

void Configuration::OnEditBoat ( wxCommandEvent& event )
{
    m_pBoatDialog->Show(!m_pBoatDialog->IsShown());
}
